### PR TITLE
Mongodb-replicaset catalog item

### PR DESCRIPTION
**Pull request by galal-hussein**
A catalog item to create MongoDB-replicaSet, it consists of:
- MongoDB service which automatically connects to replicaset of minimum 3 nodes.
- MongoDB datavolume Sidekick service to the MongoDB service.
- MongoDB initialization service of only one container to only initialize the replicaSet and then exits.

### CODE REVIEW

File: templates/mongo-replicaset/0/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  - "${replset_name}"
+  image: husseingalal/mongo-cluster
+  labels:
+    io.rancher.container.hostname_override: container_name
+    io.rancher.sidekicks: mongo-datavolume
+  stdin_open: true
+  volumes_from:
+    - mongo-datavolume
+mongo-datavolume:
+  labels:
```
Comment (cloudnautique) [inline templates/mongo-replicaset/0/docker-compose.yml]: Can you make this net: none

File: templates/mongo-replicaset/0/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+mongo-init:
+  restart: 'no'
```
Comment (cloudnautique) [inline templates/mongo-replicaset/0/docker-compose.yml]: Can this be made into a sidekick? Having a service completely stopped, makes the whole stack go into a degraded mode. If you only need to run it once, and it needs to be run by only one container, you can copy what we did for glusterfs.

[Elect a leader](https://github.com/rancher/compose-templates/blob/master/glusterfs/containers/0.1.3/glusterfs/lowest_idx.sh)
[Leader script](https://github.com/rancher/compose-templates/blob/master/glusterfs/containers/0.1.3/glusterfs/replicated_volume_create.sh#L22)

**Comment by cloudnautique** [comment]
Any chance the image can be an SVG? Not a hard requirement, but a preference.

Also, what would it take to make it so you could scale up? and secondarily down. 
Also, soft anti-affinity rules would be nice so it will try to keep them separate.

**Comment by cloudnautique** [comment]
Also, for the images, can you try and use the upstream image directly, and create an Alpine linux sidekick with your scripts/etc? Take a look at the elasticsearch templates for the approach: [es templates](https://github.com/rancher/compose-templates/blob/master/elasticsearch/0.3.1/docker-compose.yml)

The reason to do this, is you don't have to build a new image for every version of Mongo, you would only have to update the compose file.

**Comment by galal-hussein** [comment]
@cloudnautique I removed the mongo-init container and used [lowest_idx.sh](https://github.com/rancher/compose-templates/blob/master/glusterfs/containers/0.1.3/glusterfs/lowest_idx.sh) with mongo-cluster image, it worked, thanks!

The image already designed to scale up  automatically, the image connects to the master and adds itself to the replicaset, and in case of the scaledown i think Mongo handles this situation itself.

**Comment by cloudnautique** [comment]
@galal-hussein awesome! The experience for that will be a lot better. 

I had tried to scale it up and it failed, thats why I mentioned it. When I listed the hosts it only gave me the original three. 

Is it possible to use the official MongoDB image from Docker hub and put all your scripts into a sidekick?

**Comment by galal-hussein** [comment]
@cloudnautique I tweaked the image and the scaling is working now.

Regarding using the upstream image, i can't figure out how this will work in my case, because i am using python scripts to (scale, initiate the cluster) and i installed different libraries, correct me if i am wrong but if i used it as a sidekick i won't be able to run scripts properly then.

**Comment by cloudnautique** [comment]
It looks like your talking over the network? To do the clustering? You can launch the sidekick in in `net: "container:mongo"` and run it from the sidekick directly. If that doesn't work, could you create a virtualenv in a volume and launch from there? might be a little tricky due to libs...

WIth the volumes_from approach you override the upstream image entrypoint with the one you plan to inject.

**Comment by galal-hussein** [comment]
@cloudnautique I think i did it, i used bash scripts instead of python scripts in the [mongo-cluster](https://github.com/galal-hussein/rancher-mongo-config) image, and used it as a sidekick to the upstream image, also tested the scaling with this setup.

File: templates/mongo-replicaset/0/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  - "${replset_name}"
+  image: mongo:3.0
+  labels:
+    io.rancher.container.hostname_override: container_name
+    io.rancher.sidekicks: mongo-base, mongo-datavolume
+  volumes_from:
+    - mongo-datavolume
+    - mongo-base
+mongo-base:
+  restart: always
```
Comment (cloudnautique) [inline templates/mongo-replicaset/0/docker-compose.yml]: Can this be net: none or net container?

File: templates/mongo-replicaset/0/rancher-compose.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+.catalog:
+  name: "MongoDB-replicaSet"
+  version: "3.0.0-rancher1"
+  description: "MongoDB ReplicaSet"
+  uuid: mongodb-0
+  questions:
```
Comment (cloudnautique) [inline templates/mongo-replicaset/0/rancher-compose.yml]: Please add:
minimum_rancher_version: v0.46.0

File: templates/mongo-replicaset/catalogIcon-mongodb.svg [hunk 5]
```diff
@@ -0,0 +1,161 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (cloudnautique) [inline templates/mongo-replicaset/catalogIcon-mongodb.svg]: The SVG ends up not fitting in the space allotted.
screen shot 2015-11-23 at 10 55 49 am

File: templates/mongo-replicaset/0/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  tty: true
+  labels:
+    io.rancher.container.hostname_override: container_name
+    io.rancher.container.start_once: true
+  image: husseingalal/mongo-cluster:latest
```
Comment (cloudnautique) [inline templates/mongo-replicaset/0/docker-compose.yml]: Can you open a PR to rancher/compose-templates in the appropriate format? We will need to build this image and host it in the Rancher repo.

**Comment by cloudnautique** [comment]
Can we just call the folder/name of the app MongoDB?

**Comment by cloudnautique** [comment]
Looking good. I was able to verify that it deployed and scaled up.

**Comment by galal-hussein** [comment]
@cloudnautique Done, i will also open a PR to  rancher/compose-templates.

File: templates/MongoDB/0/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    - mongo-datavolume
+    - mongo-base
+mongo-base:
+  restart: always
+  net: none
+  tty: true
+  labels:
+    io.rancher.container.hostname_override: container_name
+    io.rancher.container.start_once: true
+  image: husseingalal/mongo-cluster:latest
```
Comment (cloudnautique) [inline templates/MongoDB/0/docker-compose.yml]: Rancher image is pushed.

**Comment by galal-hussein** [comment]
Tested and worked

**Comment by cloudnautique** [comment]
OK, this looks good. Can you squash and rebase?

**Comment by galal-hussein** [comment]
@cloudnautique done.

**Comment by cloudnautique** [comment]
LGTM, @vincent99 can you do a once over please.

**Comment by vincent99** [comment]
I would think name should be "MongoDB" and the description something like "MongoDB replication set cluster".. it does not appear that they use the spelling/case `replicaSet` anywhere in documentation.

Unless there have specific logo usage guidelines that would be violated, I would prefer if the logo were just the leaf graphic like (most of) the other templates, since the name is already part of the template name and description.

**Comment by galal-hussein** [comment]
@vincent99 I believe its called Replica Set, so how about the name just "MongoDB" and the description "MongoDB Replica Set".

I think they have a [guideline](https://www.mongodb.com/brand-resources) for not distorting the MongoDB logo.

**Comment by vincent99** [comment]
Distorting is clearly referring to non-proportionally scaling the image.. I don't see anything in their guideline to suggest we can't use just the leaf (http://vtf.me/mongodb-leaf.svg).

**Comment by galal-hussein** [comment]
@vincent99 modifed and commits squashed.

**Comment by cloudnautique** [comment]
LGTM

**Comment by jshimko** [comment]
Just curious... how is the keyfile being generated for the [internal auth](https://docs.mongodb.org/manual/tutorial/enable-internal-authentication/) that replica sets use?  Additionally, how is that being handled when scaling up happens?  (every member needs the same key)

And to your comment about scaling down, I don't think that's automatic for Mongo.  Pretty sure you have to update the replica set config on the primary to remove the member (see [rs.remove()](https://docs.mongodb.org/manual/reference/method/rs.remove/#rs.remove)).  I don't think it will break anything if you don't do that step, but the set will be assuming that the missing member is only temporarily offline.  And then what happens if you scale up again later?  Is the same config re-used or does it add _another_ member to that config.  If the latter, that clearly makes for a confusing/broken replica set configuration under the hood.  Every time you scale up and back down you'd end up with another unused member remaining in your config. 

Forgive me if this is an obvious answer, but where is the code that's doing all of the initialization?  All I see is a custom entrypoint script being used on the official Mongo container.  Also, where does that live? (this: `/opt/rancher/bin/entrypoint.sh`)

Sorry, just getting up to speed on the new catalog stuff.  Awesome work, by the way!

**Comment by galal-hussein** [comment]
@jshimko Enabling internal auth not handled currently by the initiation script and it may be handled by entering the key as a value to question when building the catalog item, as for the second point i think handling such cases automatically will require another container to check for the offline members and remove them regularly from the configuration and may cause a lot false positives, IMHO it should be done manually.

You can find the Mongo compose template [here](https://github.com/rancher/compose-templates/tree/master/MongoDB/containers/0.1.0/mongodb-config), which contains the initialization scripts.

**Comment by jshimko** [comment]
Thanks for the quick response @galal-hussein.  

I have a container I use to build/run an automated replica set with internal auth, user/db creation, etc.  However, it doesn't have the extra data containers or scaling features (always assumes 3 containers).

My goal is to have all of the features I currently have plus all of the features you built here.  I'll keep you posted as I make progress on that.  Most of the hard stuff is done, so it shouldn't be too bad (I know, famous last words).

Thanks for the info!

**Comment by smw355** [comment]
We are thinking about how to make an open public marketplace of catalog items.  It would be great if you can share your final version when you get tit finished.

**Pull request merged**